source/Vignette/Rendering/Effect.cs

```diff
@@ -0,0 +1,223 @@
+    /// Creates an <see cref="Effect"/> from a <see cref="ReadOnlySpan{byte}"/>.
+    /// </summary>
+    /// <param name="bytes">The bytes to be read.</param>
+    /// <param name="encoding">The text encoding.</param>
+    /// <returns>An effect.</returns>
+    public static Effect From(ReadOnlySpan<byte> bytes, Encoding? encoding = null)
+    {
+        return From((encoding ?? Encoding.UTF8).GetString(bytes));
+    }
+
+    /// <summary>
+    /// Creates an <see cref="Effect"/> from a <see langword="string"/>.
+    /// </summary>
+    /// <param name="text">The text to be compiled.</param>
+    /// <returns>An effect.</returns>
+    public static Effect From(string text)
+    {
+        string code = sh_common + text;
+
+        var shVert = ShaderCode.From(code, ShaderStage.Vertex, sh_vert, ShaderLanguage.HLSL);
+        var shFrag = ShaderCode.From(code, ShaderStage.Fragment, sh_frag, ShaderLanguage.HLSL);
+
+        var shVertReflect = shVert.Reflect();
+        var shFragReflect = shFrag.Reflect();
+
+        var parameters = new Dictionary<string, IParameter>();
+
+        if (shVertReflect.Uniforms is not null)
+        {
+            foreach (var buffer in shVertReflect.Uniforms)
+                parameters[buffer.Name] = new Parameter<GraphicsBuffer>(buffer.Name, buffer.Binding);
+        }
+
+        if (shVertReflect.Textures is not null)
+        {
+            foreach (var texture in shVertReflect.Textures)
+                parameters[texture.Name] = new Parameter<Texture>(texture.Name, texture.Binding);
+        }
+
+        if (shFragReflect.Uniforms is not null)
+        {
+            foreach (var buffer in shFragReflect.Uniforms)
+                parameters[buffer.Name] = new Parameter<GraphicsBuffer>(buffer.Name, buffer.Binding);
+        }
+
+        if (shFragReflect.Textures is not null)
+        {
+            foreach (var texture in shFragReflect.Textures)
+                parameters[texture.Name] = new Parameter<Texture>(texture.Name, texture.Binding);
+        }
+
+        return new(shVert, shFrag, parameters);
+    }
+
+    public bool Equals(Effect other)
+    {
+        return ((IStructuralEquatable)shaderCode).Equals(other.shaderCode, EqualityComparer<ShaderCode>.Default);
+    }
+
+    public override bool Equals([NotNullWhen(true)] object? obj)
+    {
+        return obj is Effect effect && Equals(effect);
+    }
+
+    public override int GetHashCode()
+    {
+        return ((IStructuralEquatable)shaderCode).GetHashCode(EqualityComparer<ShaderCode>.Default);
+    }
+
+    public static bool operator ==(Effect left, Effect right)
+    {
+        return left.Equals(right);
+    }
+
+    public static bool operator !=(Effect left, Effect right)
+    {
+        return !(left == right);
+    }
+
+    public static implicit operator ShaderCode[](Effect effect) => effect.shaderCode;
+
+    private const string sh_frag = "Pixel";
+    private const string sh_vert = "Vertex";
+
+    private const string sh_common =
```

Comment:
For brevity, can we turn this into a file in the future? Would be nice to have this as part of our "shader standard library" kinda like Blender.

Reply:
I'll leave this as a task when we have a working content manager system going on.

Reply:
For context [Unity does allow referring Shader code](https://docs.unity3d.com/2021.1/Documentation/Manual/shader-include-directives.html#:~:text=In%20HLSL%2C%20%23include%20directives%20are,same%20as%20in%20standard%20HLSL.)